src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+    logger.info("Filtering and sorting input.")
+    in.iterator
+      .filter(r => !r.isSecondaryOrSupplementary)
+      .filter(r => r.getReadPairedFlag)
+      .filter(r => !r.getReadUnmappedFlag && !r.getMateUnmappedFlag)
+      .filter(r => r.getReferenceIndex == r.getMateReferenceIndex)
+      .filter(r => r.getMappingQuality >= this.minMapQ && r.getIntegerAttribute(SAMTag.MQ.name()) >= this.minMapQ)
+      .foreach(r => { sorter.add(r); sortProgress.record(r) })
+
+    // Output the reads in the new ordering
+    logger.info("Assigning reads to UMIs and outputting.")
+    val outHeader = header.clone()
+    outHeader.setSortOrder(SortOrder.unsorted)
+    outHeader.setGroupOrder(GroupOrder.query)
+    val out = new SAMFileWriterFactory().makeWriter(outHeader, true, output.toFile, null)
+    val outProgress = new ProgressLogger(logger, verb="Output")
+
+    val iterator = sorter.iterator().grouped(2).map(i => (i(0), i(1))).buffered // consume in pairs
+    while (iterator.hasNext) {
+      // Take the next set of pairs by position and assign UMIs
+      val pairs = takeNextGroup(iterator)
+      pairs.foreach { case(r1, r2) => assert(r1.getReadName == r2.getReadName, "Reads out of order @ " + r1.getReadName) }
+      assignUmiGroups(pairs)
+
+      // Then output the records in the right order (assigned tag, read name, r1, r2)
+      val pairsByAssignedTag = pairs.groupBy(pair => pair._1.getStringAttribute(this.assignTag))
+      pairsByAssignedTag.keys.toSeq.sorted.foreach(tag => {
+        pairsByAssignedTag(tag).sortBy(pair => pair._1.getReadName).flatMap(pair => Seq(pair._1, pair._2)).foreach(rec => {
+          out.addAlignment(rec)
+          outProgress.record(rec)
+        })
+      })
+    }
+
+    out.close()
+  }
+
+  /** Consumes the next group of pairs with all matching end positions and returns them. */
+  def takeNextGroup(iterator: BufferedIterator[ReadPair]) : Seq[ReadPair] = {
+    val first = iterator.next()
+    val firstEnds = extractEnds(first._1)
+    val buffer = ListBuffer[(SAMRecord, SAMRecord)]()
+    while (iterator.hasNext &&
+           library(first._1) == library(iterator.head._1) &&
+           firstEnds == extractEnds(iterator.head._1)) buffer += iterator.next()
+    first :: buffer.toList
+  }
+
+  /**
+    * Takes in pairs of reads and writes an attribute (specified by `assignTag`) denoting
+    * sub-grouping into UMI groups by original molecule.
+    */
+  def assignUmiGroups(pairs: Seq[ReadPair]): Unit = {
+    val rawUmis       = pairs.map(_._1).map(_.getStringAttribute(this.rawTag).toUpperCase)
+    val rawToAssigned = this.assigner.assign(rawUmis)
+    val assignedToId  = rawToAssigned.values.toSet[String].map(_ -> counter.getAndIncrement().toString).toMap
+
+    pairs.foreach(pair => Seq(pair._1, pair._2).foreach(rec => {
+      val raw = rec.getStringAttribute(this.rawTag).toUpperCase
+      val assigned = rawToAssigned(raw)
+      val id = assignedToId(assigned)
```

Comment:
I'm not sure I understand.  The way it's implemented it does a double-mapping: `raw-umi` -> `assigned-umi` -> `serial-id`.  I did this so that the assignment strategies can just map input to output UMIs, and then one piece of code can "name" each group.

One question is whether we would prefer to use more meaningful IDs, e.g. instead of `1`, maybe something like `chr1:1000-1400:+-:ACTGCATG`.  The latter is clearly more human meaningful, but to guaranteed to be unique we'd have to include a bunch of fields.